### PR TITLE
Add sticky session support to containous/oxy's round robin

### PR DESCRIPTION
**Pull request by owen**
Bonjour @emilevauge,

This change is the same I've pushed through to vulcand/oxy(https://github.com/vulcand/oxy/pull/41) - but that hasn't been accepted, so I'll throw it over here!

I've got another PR that depends on this one for Traefik to use it, that will be coming soon...

I've not implemented this PR for the rebalancing round robin since I'm not certain how that will function... any comments are appreciated :)

### CODE REVIEW

**Comment by emilevauge** [comment]
Hi @owen, thanks a lot!
Even if I don't promote the use of stickysessions, this is a great PR ;)
One question: what are the difficulties implementing this in rebalancing round robin ?

**Comment by owen** [comment]
I don't like sticky sessions either.  But we have ~100 legacy applications that I don't want to rewrite 😆 

> One question: what are the difficulties implementing this in rebalancing round robin ?

There shouldn't be a technical concern, I just haven't decided the best way to handle it -- I _think_ what would be best is for rebalancing to occur for new connections, but for connections that have a cookie we keep the same backend.

I'll probably work on that today, actually!

**Comment by emilevauge** [comment]
Great!

>  I think what would be best is for rebalancing to occur for new connections, but for connections that have a cookie we keep the same backend.

Seems good :)

**Comment by owen** [comment]
OK - the logic remains unchanged for new connections; if, however, a cookie is set the connection will be stuck.

I'm going to ping the vulcand maintainers to see if they are interested at all as well

**Comment by emilevauge** [comment]
Thanks a lot @owen, great job!
LGTM
If https://github.com/vulcand/oxy/pull/41 has not been merged in a few days, I will merge this one as is.

File: roundrobin/rr.go [hunk 1]
```diff
@@ -62,14 +71,35 @@ func (r *RoundRobin) Next() http.Handler {
...
 func (r *RoundRobin) ServeHTTP(w http.ResponseWriter, req *http.Request) {
-	url, err := r.NextServer()
-	if err != nil {
-		r.errHandler.ServeHTTP(w, req, err)
-		return
-	}
 	// make shallow copy of request before chaning anything to avoid side effects
 	newReq := *req
-	newReq.URL = url
+	stuck := false
```
Comment (emilevauge) [inline roundrobin/rr.go:76]: Hum, there is an issue here I think.
You don't call `url, err := r.NextServer()` if `stuck` is false.

Comment (owen) [inline roundrobin/rr.go:76]: Glad I saw your comment in my other PR, I didn't see this comment!

I've retraced the logic of that section.  The flow is this:
- Look to see if we have a sticky session
  *\* if so, pull the server out of the cookie and store it.
  *\* if not, fall through to the next session
- Look to see if we use sticky sessions & do not have a server || if we do not use sticky sessions:
  *\* call NextServer() to get the approp. URL
  *\* if we have sticky sessions, stick that backend
- Return the calculated backend URL

So we only call NextServer() if we have failed to retrieve a server URL from the cookie - that's in the `if !stuck {` stanza.  We won't call it if `stuck = true`, which is only if we have retrieved the server from the cookie -- and, validated that it is still a valid server, which we handle in `ss.GetBackend`.

Comment (owen) [inline roundrobin/rr.go:76]: @emilevauge let me know if that clears it up!  alternatively, if I'm missing something here - which I may be ;) - let me know and I can revise

**Comment by emilevauge** [comment]
Hi @owen, sorry for the delay... We add to wait for traefik 1.0.0 stable before touching oxy ;)
Merging this one :+1:

**Comment by emilevauge** [comment]
Hum, there is a minor issue in the test `TestBadCookieVal`, it passes alone, but fails while running the whole test suite. @owen any idea?

**Comment by owen** [comment]
That's odd. I'll look at this today!

**Pull request merged**